epochStart/metachain/rewardsV2.go

```diff
@@ -0,0 +1,496 @@
+	validatorsInfo map[uint32][]*state.ValidatorInfo,
+	metaBlock *block.MetaBlock,
+	miniBlocks block.MiniBlockSlice,
+	rewardsPerNode map[uint32][]*big.Int,
+) (*big.Int, error) {
+	rwdAddrValidatorInfo, accumulatedDust := rc.computeValidatorInfoPerRewardAddress(validatorsInfo, rewardsPerNode)
+
+	for _, rwdInfo := range rwdAddrValidatorInfo {
+		rwdTx, rwdTxHash, err := rc.createRewardFromRwdInfo(rwdInfo, metaBlock)
+		if err != nil {
+			return nil, err
+		}
+		if rwdTx.Value.Cmp(zero) <= 0 {
+			continue
+		}
+
+		rc.accumulatedRewards.Add(rc.accumulatedRewards, rwdTx.Value)
+		mbId := rc.shardCoordinator.ComputeId([]byte(rwdInfo.address))
+		if mbId == core.MetachainShardId {
+			mbId = rc.shardCoordinator.NumberOfShards()
+
+			if !rc.flagDelegationSystemSCEnabled.IsSet() || !rc.isSystemDelegationSC(rwdTx.RcvAddr) {
+				accumulatedDust.Add(accumulatedDust, rwdTx.GetValue())
+				continue
+			}
+		}
+
+		if rwdTx.Value.Cmp(big.NewInt(0)) < 0 {
+			log.Error("negative rewards", "rcv", rwdTx.RcvAddr)
+			continue
+		}
+		rc.currTxs.AddTx(rwdTxHash, rwdTx)
+		miniBlocks[mbId].TxHashes = append(miniBlocks[mbId].TxHashes, rwdTxHash)
+	}
+
+	return accumulatedDust, nil
+}
+
+func (rc *rewardsCreatorV2) computeValidatorInfoPerRewardAddress(
+	validatorsInfo map[uint32][]*state.ValidatorInfo,
+	rewardsPerNode map[uint32][]*big.Int,
+) (map[string]*rewardInfoData, *big.Int) {
+
+	rwdAddrValidatorInfo := make(map[string]*rewardInfoData)
+	accumulatedUnassigned := big.NewInt(0)
+
+	for shardID, shardValidatorsInfo := range validatorsInfo {
+		for i, validatorInfo := range shardValidatorsInfo {
+			if validatorInfo.LeaderSuccess == 0 && validatorInfo.ValidatorSuccess == 0 {
+				accumulatedUnassigned.Add(accumulatedUnassigned, rewardsPerNode[shardID][i])
+				continue
+			}
+
+			rwdInfo, ok := rwdAddrValidatorInfo[string(validatorInfo.RewardAddress)]
+			if !ok {
+				rwdInfo = &rewardInfoData{
+					accumulatedFees: big.NewInt(0),
+					protocolRewards: big.NewInt(0),
+					address:         string(validatorInfo.RewardAddress),
+				}
+				rwdAddrValidatorInfo[string(validatorInfo.RewardAddress)] = rwdInfo
+			}
+
+			rwdInfo.accumulatedFees.Add(rwdInfo.accumulatedFees, validatorInfo.AccumulatedFees)
+			rwdInfo.protocolRewards.Add(rwdInfo.protocolRewards, rewardsPerNode[shardID][i])
+		}
+	}
+
+	return rwdAddrValidatorInfo, accumulatedUnassigned
+}
+
+// IsInterfaceNil return true if underlying object is nil
+func (rc *rewardsCreatorV2) IsInterfaceNil() bool {
+	return rc == nil
+}
+
+func (rc *rewardsCreatorV2) prepareStakingData(eligibleNodesKeys map[uint32][][]byte) error {
+	sw := core.NewStopWatch()
+	sw.Start("prepareRewardsFromStakingSC")
+	defer func() {
+		sw.Stop("prepareRewardsFromStakingSC")
+		log.Debug("rewardsCreator.prepareRewardsFromStakingSC time measurements", sw.GetMeasurements())
+	}()
+
+	return rc.stakingDataProvider.PrepareStakingData(eligibleNodesKeys)
+}
+
+func (rc *rewardsCreatorV2) getEligibleNodesKeyMap(
```

Comment:
this is actually all validators.

Reply:
good find, changed to consider only validators from eligible list.

Reply:
but don;t we need some other validators from the last blocks of currentEpoch - 2. Those which were notarized by meta in epoch -1 ?

Reply:
as discussed, this will simplify the computation.